tools/scripts/update-bumpenvs-yaml.py

```diff
@@ -0,0 +1,169 @@
+
+Usage: update-bumpenvs-yaml.py path/to/bumpenvs.yaml ENVIRONMENTS_COMMIT
+"""
+
+import requests
+import os
+import sys
+import yaml
+import io
+from typing import Any, Dict
+
+USER = "determined-ai"
+PROJECT = "environments"
+BASE_URL = f"https://circleci.com/api/v1.1/project/github/{USER}/{PROJECT}"
+
+EXPECT_JOBS = {
+    "publish-cloud-images",
+    "build-and-publish-docker-tf2-cpu",
+    "build-and-publish-docker-tf1-cpu",
+    "build-and-publish-docker-tf1-gpu",
+    "build-and-publish-docker-tf2-gpu",
+}
+
+EXPECT_ARTIFACTS = {
+    "packer-log",
+    "publish-tf2-cpu",
+    "publish-tf1-cpu",
+    "publish-tf1-gpu",
+    "publish-tf2-gpu",
+}
+
+
+class Build:
+    """A neatly parsed CircleCI build."""
+
+    def __init__(self, data: Dict["str", Any]) -> None:
+        self.build_num = data["build_num"]
+        self.job_name = data["workflows"]["job_name"]
+        assert (
+            data["status"] == "success"
+        ), f"'{self.job_name}' ({self.build_num}) failed"
+
+    def get_artifacts(self) -> Dict[str, str]:
+        print(f"fetching artifacts for {self.job_name}", file=sys.stderr)
+        url = f"{BASE_URL}/{self.build_num}/artifacts"
+        req = requests.get(url)
+        req.raise_for_status()
+
+        artifacts = {}
+        for meta in req.json():
+            path = os.path.basename(meta["path"])
+            print(f"fetching {path}", file=sys.stderr)
+            req = requests.get(meta["url"])
+            req.raise_for_status()
+            artifacts[path] = req.content.decode("utf8")
+        return artifacts
+
+
+def get_all_builds(commit: str) -> Dict[str, Build]:
+    # Get all the recent jobs.
+    print("fetching recent jobs", file=sys.stderr)
+    req = requests.get(BASE_URL, params={"limit": 50, "filter": "completed"})  # type: ignore
+    req.raise_for_status()
+
+    # Get all the build numbers matching this commit.
+    builds = {}
+    for build_meta in req.json():
+        if build_meta["vcs_revision"] == commit:
+            build = Build(build_meta)
+            builds[build.job_name] = build
+
+    found = set(builds.keys())
+    assert EXPECT_JOBS == found, f"expected jobs ({EXPECT_JOBS}) but found ({found})"
+
+    return builds
+
+
+def get_all_artifacts(builds: Dict[str, Build]) -> Dict[str, str]:
+    artifacts = {}
+    for b in builds.values():
+        artifacts.update(b.get_artifacts())
+
+    found = set(artifacts.keys())
+    assert (
+        EXPECT_ARTIFACTS == found
+    ), f"expected artifacts ({EXPECT_ARTIFACTS}) but found ({found})"
+
+    return artifacts
+
+
+def parse_packer_log(packer_log: str) -> Dict[str, str]:
+    """Parse the output of packer's -machine-readable format, strange though it may be."""
+    out = {}
+
+    lines = packer_log.strip().split("\n")
+    fields = [line.split(",") for line in lines]
+    # We actually don't care about non-artifact lines or lines with fewer than 6 fields.
+    fields = [f for f in fields if len(f) > 5 and f[2] == "artifact"]
```

Comment:
non-blocking: Assuming we care only about lines with _exactly_ 6 fields, could this use `namedtuple`/`typing.NamedTuple` instead? There are way too many magic indices going around. From a quick look at the output format, I get the impression that even more structure could be added (looks like field 3 is some kind of index, which can be used to group fields 4 and 5, which are attribute name/value pairs), but I don't know if it's worth going that far. Could be nice, though.

Reply:
I took the `collections.namedtuple` option here.